src/main/java/fi/otavanopisto/kuntaapi/server/rest/ServicesApiImpl.java

```diff
@@ -0,0 +1,510 @@
+    if (serviceId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_ID, serviceIdParam));
+    }
+    
+    PrintableFormChannelId printableFormChannelId = toPrintableFormChannelId(printableFormChannelIdParam);
+    if (printableFormChannelId == null) {
+      return createBadRequest(String.format(INVALID_PRINTABLE_FORM_CHANNEL_ID, serviceIdParam));
+    }
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      PrintableFormChannel printableFormChannel = serviceChannelProvider.findPrintableFormChannel(serviceId, printableFormChannelId);
+      if (printableFormChannel != null) {
+        return Response.ok(printableFormChannel)
+            .build();
+      }
+    }
+
+    return Response
+      .status(Status.NOT_FOUND)
+      .entity(NOT_FOUND)
+      .build();
+  }
+
+  @Override
+  public Response findServiceServiceLocationChannel(String serviceIdParam, String serviceLocationChannelIdParam) {
+    ServiceId serviceId = toServiceId(serviceIdParam);
+    if (serviceId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_ID, serviceIdParam));
+    }
+    
+    ServiceLocationChannelId serviceLocationChannelId = toServiceLocationChannelId(serviceLocationChannelIdParam);
+    if (serviceLocationChannelId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_LOCATION_CHANNEL_ID, serviceIdParam));
+    }
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      ServiceLocationChannel serviceLocationChannel = serviceChannelProvider.findServiceLocationChannel(serviceId, serviceLocationChannelId);
+      if (serviceLocationChannel != null) {
+        return Response.ok(serviceLocationChannel)
+            .build();
+      }
+    }
+    
+    return Response
+      .status(Status.NOT_FOUND)
+      .entity(NOT_FOUND)
+      .build();
+  }
+
+  @Override
+  public Response findServiceWebPageChannel(String serviceIdParam, String webPageChannelIdParam) {
+    ServiceId serviceId = toServiceId(serviceIdParam);
+    if (serviceId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_ID, serviceIdParam));
+    }
+    
+    WebPageChannelId webPageChannelId = toWebPageChannelId(webPageChannelIdParam);
+    if (webPageChannelId == null) {
+      return createBadRequest(String.format(INVALID_WEBPAGE_CHANNEL_ID, serviceIdParam));
+    }
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      WebPageChannel webPageChannel = serviceChannelProvider.findWebPageChannelChannel(serviceId, webPageChannelId);
+      if (webPageChannel != null) {
+        return Response.ok(webPageChannel)
+            .build();
+      }
+    }
+    
+    return Response
+      .status(Status.NOT_FOUND)
+      .entity(NOT_FOUND)
+      .build();
+  }
+
+  @Override
+  public Response listServiceElectronicChannels(String serviceIdParam, Long firstResult, Long maxResults) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "listServiceElectronicChannels" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)